src/hooks/useValidatedSearchParams.ts

```diff
@@ -0,0 +1,252 @@
+import { contentSearchValidators, resourceSearchValidators } from "./validation"
+import type { QueryParamValidators } from "./validation"
+import type { v1 } from "@mitodl/open-api-axios"
+import { useCallback, useMemo, useState } from "react"
+
+type ResourceSearchRequest =
+  v1.LearningResourcesSearchApiLearningResourcesSearchRetrieveRequest
+type ContentFileSearchRequest =
+  v1.ContentFileSearchApiContentFileSearchRetrieveRequest
+
+interface UseValidatedSearchParamsProps<ReqParams> {
+  /**
+   * Source of truth for search parameters.
+   *
+   * Note: React should be aware of the state of this object. For example,
+   * do NOT pass `new UrlSearchParams(window.location.search)` directly.
+   * Instead, use `useSearchParams` hook from `react-router-dom` or equivalent.
+   */
+  searchParams: URLSearchParams
+  /**
+   * A setter for `searchParams`.
+   */
+  setSearchParams: (
+    value: URLSearchParams | ((prev: URLSearchParams) => URLSearchParams)
+  ) => void
+  /**
+   * Array of request parameter keys that are considered facets.
+   * Determs when to call `onFacetsChange` callback and which params are reset
+   * via clearAllFacets.
+   */
+  facets?: (keyof ReqParams & string)[]
+  onFacetsChange?: () => void
+}
+
+interface UseValidatedSearchParamsResult<ReqParams> {
+  /**
+   * Valdiated search request parameters.
+   */
+  params: ReqParams
+  /**
+   * Clear all parameters marked as facets.
+   */
+  clearAllFacets: () => void
+  /**
+   * Clear a particle parameter by key.
+   */
+  clearParam: (key: keyof ReqParams & string) => void
+  /**
+   * Patch the request parameters.
+   * - missing keys are ignored
+   * - null or empty values are removed
+   * - all other values are set (NOT appended).
+   */
+  patchParams: (patch: Partial<ReqParams>) => void
+  /**
+   * Toggle a parameter value on or off.
+   * - if `checked=true`, value is APPENDED to parameter list.
+   * - if `checked=false`, value is REMOVED from parameter list.
+   */
+  toggleParamValue: (name: string, rawValue: string, checked: boolean) => void
+  /**
+   * Current search text. May be different from the `q` parameter.
+   */
+  currentText: string
+  /**
+   * Set current text.
+   */
+  setCurrentText: (value: string) => void
+  /**
+   * Set current text and update the `q` parameter.
+   */
+  setCurrentTextAndQuery: (value: string) => void
+}
+
+const useValidatedSearchParams = <ReqParams>({
+  searchParams,
+  setSearchParams,
+  facets = [],
+  validators,
+  onFacetsChange
+}: UseValidatedSearchParamsProps<ReqParams> & {
+  validators: QueryParamValidators<ReqParams>
+}): UseValidatedSearchParamsResult<ReqParams> => {
```

Comment:
There are some notable differences in these hooks vs the `useSearchQueryParams`, which these two hooks replace.

1. **"Facets" are less special**: The old hook returned a `params` object with shape `{ queryText, sort, activeFacets }`. Everything was a facet *except* queryText and `sort`. 

    This was problematic for some UIs, e.g., [this one](https://www.figma.com/proto/bRb19HaucO5loHGSU81nFG/MIT_Prototype?page-id=0%3A1&type=design&node-id=9-6102&viewport=4023%2C1145%2C0.25&t=6VyDzTtHVzsZ3Nyv-1&scaling=scale-down&starting-point-node-id=48%3A2446&show-proto-sidebar=1) which has (A) resource tbas and (B) a sidebar for facets. Previously, the "Clear All" button would reset the resource tab, even though it's not part of the filtering sidebar. 

   In the new hooks, facets are determined by a parameter, `facets`, and `clearAllFacets` only clears the parameters that are listed as `facets`. 

2. **Two separate hooks for ** for `/api/v1/learning_resources_search` and `/api/v1/content_files_search`. This makes the typescript *much* simpler, and seems reasonable since the APIs are separate.

3. Uses API parameter names, not abbreviations. ("department" not "d").